commit 3389cae427b4032c3a991cb32b5178a85a652d84
Author: Stefan Liebler <stli@linux.ibm.com>
Date:   Tue Dec 18 13:57:20 2018 +0100

    S390: Refactor wcscat ifunc handling.
    
    The ifunc handling for wcscat is adjusted in order to omit ifunc
    if the minimum architecture level already supports newer CPUs by default.
    Unfortunately the c ifunc variant can't be omitted at all as it is used
    by the z13 ifunc variant as fallback if the pointers are not 4-byte aligned.
    
    ChangeLog:
    
            * sysdeps/s390/multiarch/Makefile
            (sysdep_routines): Remove wcscat variants.
            * sysdeps/s390/Makefile (sysdep_routines): Add wcscat variants.
            * sysdeps/s390/multiarch/ifunc-impl-list.c
            (__libc_ifunc_impl_list): Refactor ifunc handling for wcscat.
            * sysdeps/s390/multiarch/wcscat-c.c: Move to ...
            * sysdeps/s390/wcscat-c.c: ... here and adjust ifunc handling.
            * sysdeps/s390/multiarch/wcscat-vx.S: Move to ...
            * sysdeps/s390/wcscat-vx.S: ... here and adjust ifunc handling.
            * sysdeps/s390/multiarch/wcscat.c: Move to ...
            * sysdeps/s390/wcscat.c: ... here and adjust ifunc handling.
            * sysdeps/s390/ifunc-wcscat.h: New file.

diff --git a/sysdeps/s390/Makefile b/sysdeps/s390/Makefile
index 640177370382235f..5fd00ad4a8661c4c 100644
--- a/sysdeps/s390/Makefile
+++ b/sysdeps/s390/Makefile
@@ -65,5 +65,6 @@ sysdep_routines += wcslen wcslen-vx wcslen-c \
 		   wcscpy wcscpy-vx wcscpy-c \
 		   wcpcpy wcpcpy-vx wcpcpy-c \
 		   wcsncpy wcsncpy-vx wcsncpy-c \
-		   wcpncpy wcpncpy-vx wcpncpy-c
+		   wcpncpy wcpncpy-vx wcpncpy-c \
+		   wcscat wcscat-vx wcscat-c
 endif
diff --git a/sysdeps/s390/ifunc-wcscat.h b/sysdeps/s390/ifunc-wcscat.h
new file mode 100644
index 0000000000000000..fecae21403e4c664
--- /dev/null
+++ b/sysdeps/s390/ifunc-wcscat.h
@@ -0,0 +1,53 @@
+/* wcscat variant information on S/390 version.
+   Copyright (C) 2018 Free Software Foundation, Inc.
+   This file is part of the GNU C Library.
+
+   The GNU C Library is free software; you can redistribute it and/or
+   modify it under the terms of the GNU Lesser General Public
+   License as published by the Free Software Foundation; either
+   version 2.1 of the License, or (at your option) any later version.
+
+   The GNU C Library is distributed in the hope that it will be useful,
+   but WITHOUT ANY WARRANTY; without even the implied warranty of
+   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+   Lesser General Public License for more details.
+
+   You should have received a copy of the GNU Lesser General Public
+   License along with the GNU C Library; if not, see
+   <http://www.gnu.org/licenses/>.  */
+
+#if defined USE_MULTIARCH && IS_IN (libc)		\
+  && ! defined HAVE_S390_MIN_Z13_ZARCH_ASM_SUPPORT
+# define HAVE_WCSCAT_IFUNC	1
+#else
+# define HAVE_WCSCAT_IFUNC	0
+#endif
+
+#ifdef HAVE_S390_VX_ASM_SUPPORT
+# define HAVE_WCSCAT_IFUNC_AND_VX_SUPPORT HAVE_WCSCAT_IFUNC
+#else
+# define HAVE_WCSCAT_IFUNC_AND_VX_SUPPORT 0
+#endif
+
+#if defined HAVE_S390_MIN_Z13_ZARCH_ASM_SUPPORT
+# define WCSCAT_DEFAULT		WCSCAT_Z13
+/* The z13 ifunc variant is using the common code variant as fallback!  */
+# define HAVE_WCSCAT_C		1
+# define HAVE_WCSCAT_Z13	1
+#else
+# define WCSCAT_DEFAULT		WCSCAT_C
+# define HAVE_WCSCAT_C		1
+# define HAVE_WCSCAT_Z13	HAVE_WCSCAT_IFUNC_AND_VX_SUPPORT
+#endif
+
+#if HAVE_WCSCAT_C
+# define WCSCAT_C		__wcscat_c
+#else
+# define WCSCAT_C		NULL
+#endif
+
+#if HAVE_WCSCAT_Z13
+# define WCSCAT_Z13		__wcscat_vx
+#else
+# define WCSCAT_Z13		NULL
+#endif
diff --git a/sysdeps/s390/multiarch/Makefile b/sysdeps/s390/multiarch/Makefile
index 158fb495523438b4..617017496c79e2ca 100644
--- a/sysdeps/s390/multiarch/Makefile
+++ b/sysdeps/s390/multiarch/Makefile
@@ -1,6 +1,5 @@
 ifeq ($(subdir),wcsmbs)
-sysdep_routines += wcscat wcscat-vx wcscat-c \
-		   wcsncat wcsncat-vx wcsncat-c \
+sysdep_routines += wcsncat wcsncat-vx wcsncat-c \
 		   wcscmp wcscmp-vx wcscmp-c \
 		   wcsncmp wcsncmp-vx wcsncmp-c \
 		   wcschr wcschr-vx wcschr-c \
diff --git a/sysdeps/s390/multiarch/ifunc-impl-list.c b/sysdeps/s390/multiarch/ifunc-impl-list.c
index e60238fcde4dcd4f..b05bd35fd898d0a6 100644
--- a/sysdeps/s390/multiarch/ifunc-impl-list.c
+++ b/sysdeps/s390/multiarch/ifunc-impl-list.c
@@ -52,6 +52,7 @@
 #include <ifunc-wcpcpy.h>
 #include <ifunc-wcsncpy.h>
 #include <ifunc-wcpncpy.h>
+#include <ifunc-wcscat.h>
 
 /* Maximum number of IFUNC implementations.  */
 #define MAX_IFUNC	3
@@ -502,6 +503,18 @@ __libc_ifunc_impl_list (const char *name, struct libc_ifunc_impl *array,
 		)
 #endif /* HAVE_WCPNCPY_IFUNC  */
 
+#if HAVE_WCSCAT_IFUNC
+    IFUNC_IMPL (i, name, wcscat,
+# if HAVE_WCSCAT_Z13
+		IFUNC_IMPL_ADD (array, i, wcscat,
+				dl_hwcap & HWCAP_S390_VX, WCSCAT_Z13)
+# endif
+# if HAVE_WCSCAT_C
+		IFUNC_IMPL_ADD (array, i, wcscat, 1, WCSCAT_C)
+# endif
+		)
+#endif /* HAVE_WCSCAT_IFUNC  */
+
 #ifdef HAVE_S390_VX_ASM_SUPPORT
 
 # define IFUNC_VX_IMPL(FUNC)						\
@@ -510,8 +523,6 @@ __libc_ifunc_impl_list (const char *name, struct libc_ifunc_impl *array,
 			      __##FUNC##_vx)				\
 	      IFUNC_IMPL_ADD (array, i, FUNC, 1, __##FUNC##_c))
 
-  IFUNC_VX_IMPL (wcscat);
-
   IFUNC_VX_IMPL (wcsncat);
 
   IFUNC_VX_IMPL (wcscmp);
diff --git a/sysdeps/s390/multiarch/wcscat-c.c b/sysdeps/s390/wcscat-c.c
similarity index 86%
rename from sysdeps/s390/multiarch/wcscat-c.c
rename to sysdeps/s390/wcscat-c.c
index 9a31c65a0ba668e5..bc1a50b1a52da259 100644
--- a/sysdeps/s390/multiarch/wcscat-c.c
+++ b/sysdeps/s390/wcscat-c.c
@@ -16,10 +16,12 @@
    License along with the GNU C Library; if not, see
    <http://www.gnu.org/licenses/>.  */
 
-#if defined HAVE_S390_VX_ASM_SUPPORT && IS_IN (libc)
-# define WCSCAT  __wcscat_c
+#include <ifunc-wcscat.h>
+
+#if HAVE_WCSCAT_C
+# if HAVE_WCSCAT_IFUNC || HAVE_WCSCAT_Z13
+#  define WCSCAT WCSCAT_C
+# endif
 
-# include <wchar.h>
-extern __typeof (__wcscat) __wcscat_c;
 # include <wcsmbs/wcscat.c>
 #endif
diff --git a/sysdeps/s390/multiarch/wcscat-vx.S b/sysdeps/s390/wcscat-vx.S
similarity index 96%
rename from sysdeps/s390/multiarch/wcscat-vx.S
rename to sysdeps/s390/wcscat-vx.S
index 2164a8da411eadaa..4e40d69e8417d6d1 100644
--- a/sysdeps/s390/multiarch/wcscat-vx.S
+++ b/sysdeps/s390/wcscat-vx.S
@@ -16,7 +16,8 @@
    License along with the GNU C Library; if not, see
    <http://www.gnu.org/licenses/>.  */
 
-#if defined HAVE_S390_VX_ASM_SUPPORT && IS_IN (libc)
+#include <ifunc-wcscat.h>
+#if HAVE_WCSCAT_Z13
 
 # include "sysdep.h"
 # include "asm-syntax.h"
@@ -37,7 +38,7 @@
    -v17=index of zero
    -v18=part of src
 */
-ENTRY(__wcscat_vx)
+ENTRY(WCSCAT_Z13)
 	.machine "z13"
 	.machinemode "zarch_nohighgprs"
 
@@ -170,6 +171,11 @@ ENTRY(__wcscat_vx)
 	lgr	%r2,%r0		/* Load saved dest-ptr.  */
 	br	%r14
 .Lfallback:
-	jg	__wcscat_c
-END(__wcscat_vx)
-#endif /* HAVE_S390_VX_ASM_SUPPORT && IS_IN (libc) */
+	jg	WCSCAT_C
+END(WCSCAT_Z13)
+
+# if ! HAVE_WCSCAT_IFUNC
+strong_alias (WCSCAT_Z13, __wcscat)
+weak_alias (__wcscat, wcscat)
+# endif
+#endif
diff --git a/sysdeps/s390/multiarch/wcscat.c b/sysdeps/s390/wcscat.c
similarity index 70%
rename from sysdeps/s390/multiarch/wcscat.c
rename to sysdeps/s390/wcscat.c
index 33e4f6da3ff8eea7..3741210a177f8029 100644
--- a/sysdeps/s390/multiarch/wcscat.c
+++ b/sysdeps/s390/wcscat.c
@@ -16,13 +16,24 @@
    License along with the GNU C Library; if not, see
    <http://www.gnu.org/licenses/>.  */
 
-#if defined HAVE_S390_VX_ASM_SUPPORT && IS_IN (libc)
+#include <ifunc-wcscat.h>
+
+#if HAVE_WCSCAT_IFUNC
 # include <wchar.h>
 # include <ifunc-resolve.h>
 
-s390_vx_libc_ifunc (__wcscat)
-weak_alias (__wcscat, wcscat)
+# if HAVE_WCSCAT_C
+extern __typeof (__wcscat) WCSCAT_C attribute_hidden;
+# endif
 
-#else
-# include <wcsmbs/wcscat.c>
-#endif /* !(defined HAVE_S390_VX_ASM_SUPPORT && IS_IN (libc)) */
+# if HAVE_WCSCAT_Z13
+extern __typeof (__wcscat) WCSCAT_Z13 attribute_hidden;
+# endif
+
+s390_libc_ifunc_expr (__wcscat, __wcscat,
+		      (HAVE_WCSCAT_Z13 && (hwcap & HWCAP_S390_VX))
+		      ? WCSCAT_Z13
+		      : WCSCAT_DEFAULT
+		      )
+weak_alias (__wcscat, wcscat)
+#endif
